### ISSUE TITLE
Allow file extensions to be configurable

### ISSUE THREAD

**Issue by sriv** [mkdocs/mkdocs#930]
I notice that the allowed file extensions are [in code](https://github.com/mkdocs/mkdocs/blob/3e03014894c173ce5ffaaf2a8d92ef9dfbeb865d/mkdocs/utils/__init__.py#L204).

Please consider allowing this to be configured. I have a use case where my users would author Markdown based specifications for consumption by another tool - [gauge](https://github.com/getgauge/gauge). I would like them to use mkdocs to export their markdown to html, hence the need for this feature.

**Comment by waylan** [comment]
@d0ugal you closed #924 only 3 days ago and you labeled this [Enhancment] today. Did you change your mind? To be clear, I don't care which way you rule on this, I just want to make sure I understand what your decision is.

**Comment by d0ugal** [comment]
@waylan You are right, maybe I should have marked this as design decision needed. I am a bit conflicted. I've swift it to NDD while we discuss it. Here are my thoughts, in no particular order.

- It would be quite a small change, and it is a request that seems to come up now and then...
- ... but I had adding new config option, having less is so much better.
- I didn't remember that we supported do many extensions until @sriv pointed them out. We should either support only a few or make it configurable, at the moment we are sitting in the middle.
- I think users would be wise to use a more standard extension like `.md` and if other tools don't support them, then they are clearly at fault, not us.
- Changing an existing set of documents to `.md` would be annoying as it messes with your git history... but it also isn't that bad.

**Comment by waylan** [comment]
I recently came across a few static site generators (I forget which offhand) which used an extension like `.page`. Personally, I've always preferred `.txt` for Markdown files. The creator of Markdown uses `.text` and has stated in the past that he believes Markdown should not have/need its own file extension as it is just plain text. I don't want to get into that argument here, but bring it up solely to point out that not everyone agrees on the subject. 

It appears to me that MkDocs is trying to please most everyone with the long list of Markdown file extensions it supports (much more than a large site like GitHub supports). Yet, it clearly still fails to please everyone. Therefore, I see two options:

1. Make an executive decision that MkDocs will only support a specific list of extensions and be done with it. Unfortunately, this would basically be "do nothing" as it could break existing projects if we shortened the list (my preference would be to shorten the list to `['.md', '.markdown']` if we have to have a hard-coded list, but that's not backward-compatible).

2. Make it configurable so anyone can use whatever they desire. The concern here is that if we create a setting, what do we set as the default? `.md`? Or a list? But do users really use multiple different extensions per project? If we don't want to support a list, then existing users who didn't use the default we pick will need to make a change upon upgrade -- at the least they will need to add the setting to their config file. And if we do support a list, do the entries in the config file add to the default list or replace/override the list?

> I think users would be wise to use a more standard extension like .md and if other tools don't support them, then they are clearly at fault, not us.

While I understand the sentiment here, it assumes there is a standard file extension for Markdown files, when in fact, there is not. In fact, over the years various individuals have asked the creator of Markdown to either file an application for an officially recognized extension/mime-type and/or give his blessing for them to do so. He has always refused (silence constitutes refusal here). Whether you agree with that position or not, you can't blame projects/libraries for using all sorts of non-standard file extensions for Markdown files.  Personally, I dislike it anytime I see a project/library force their preference on others.

I like `.txt` as Markdown is plain text, although I think something like `.page` is clever as it makes it more project specific (this file is a page in project X; therefore its not likely to support being exported to something else; something like `.mkd` would accomplish that). However, the extensions recognized by GitHub etc. are nice because they work in lots of places. 

I would prefer adding a setting for this, even though I agree that we should be generally reducing settings. In other words, in my opinion, this overcomes the hurdle. That said, there are still a lot of backward compatibility concerns given the already large user base. In the end, I'll accept whatever you choose @d0ugal.

**Comment by waylan** [comment]
> I think something like `.page` is clever as it makes it more project specific (this file is a page in project X; therefore its not likely to support being exported to something else; something like `.mkd` would accomplish that).

Just wanted to expound upon this by adding that MkDocs currently only supports basic Markdown syntax with the option for users to expand it via Markdown extensions. Therefore, the body of pages are for the most part portable. However, if we start to add support for things like referencing meta data and/or config settings within the document (via some sort of template syntax) or otherwise add specialized syntax, then the files become non-portable and a specialized MkDocs specific file extension would make sense. The other projects I've seen using `.page` certainly fit that description.

Of course, many of those feature requests have been punted to the future (mythical) Plugin API, which makes them a user choice. Some users could still be creating very portable documents while others may not. Therefore, a hard requirement for a MkDocs specific file extension does not make as much sense. However, offering users the option to do so may make sense if/when a plugin API becomes available.

**Comment by oxr463** [comment]
Another option to consider is implementing this in a separate plugin. That way it is still backwards-compatible, and future proof, since then anyone can use any file extension they please.

**Comment by pawamoy** [comment]
Given the difficulty of implementing this in MkDocs core, and the low interest this feature request raised (no upvotes), and, anecdotally given [gauge](https://github.com/getgauge/gauge) seems to be inactive nowadays, I'll close this as out of scope for MkDocs core.

Allowing different file extensions is part of a bigger effort to support something else than Python-Markdown to convert documents from Markdown/else to HTML. While some of us are definitely interested in bringing such flexibility to MkDocs, at its core or at least through the plugin system, the feature itself will probably not be a good fit within MkDocs core itself. We'd be very happy to see someone tackle this in the form of a plugin, and if the plugin API is not sufficient, to give us feedback so that we can work on making it more powerful/flexible :slightly_smiling_face:
